ray-operator/controllers/ray/batchscheduler/schedulermanager.go

```diff
@@ -1,93 +1,89 @@
+	schedulerinterface "github.com/ray-project/kuberay/ray-operator/controllers/ray/batchscheduler/interface"
 	"github.com/ray-project/kuberay/ray-operator/controllers/ray/utils"
 )
 
-var schedulerContainers = map[string]schedulerinterface.BatchSchedulerFactory{
-	schedulerinterface.GetDefaultPluginName(): &schedulerinterface.DefaultBatchSchedulerFactory{},
-	volcano.GetPluginName():                   &volcano.VolcanoBatchSchedulerFactory{},
-	yunikorn.GetPluginName():                  &yunikorn.YuniKornSchedulerFactory{},
+type SchedulerManager struct {
+	config     *rest.Config
+	factory    schedulerinterface.BatchSchedulerFactory
+	scheduler  schedulerinterface.BatchScheduler
+	rayConfigs configapi.Configuration
+	sync.Mutex
 }
 
-func GetRegisteredNames() []string {
-	var pluginNames []string
-	for key := range schedulerContainers {
-		pluginNames = append(pluginNames, key)
-	}
-	return pluginNames
-}
+// NewSchedulerManager maintains a specific scheduler plugin based on config
+func NewSchedulerManager(rayConfigs configapi.Configuration, config *rest.Config) (*SchedulerManager, error) {
+	var factory schedulerinterface.BatchSchedulerFactory
 
-func ConfigureReconciler(b *builder.Builder) *builder.Builder {
-	for _, factory := range schedulerContainers {
-		b = factory.ConfigureReconciler(b)
+	// init with the default implementation
+	factory = &schedulerinterface.DefaultBatchSchedulerFactory{}
+	// when a batch scheduler name is provided
+	if len(rayConfigs.BatchScheduler) > 0 {
+		switch rayConfigs.BatchScheduler {
+		case volcano.GetPluginName():
+			factory = &volcano.VolcanoBatchSchedulerFactory{}
+		case yunikorn.GetPluginName():
+			factory = &yunikorn.YuniKornSchedulerFactory{}
+		default:
+			factory = &schedulerinterface.DefaultBatchSchedulerFactory{}
+		}
 	}
-	return b
-}
 
-func AddToScheme(scheme *runtime.Scheme) {
-	for _, factory := range schedulerContainers {
-		factory.AddToScheme(scheme)
+	// legacy option, if this is enabled, register volcano
+	if rayConfigs.EnableBatchScheduler {
+		factory = &volcano.VolcanoBatchSchedulerFactory{}
 	}
-}
 
-type SchedulerManager struct {
-	config  *rest.Config
-	plugins map[string]schedulerinterface.BatchScheduler
-	sync.Mutex
-}
+	scheduler, err := factory.New(config)
+	if err != nil {
+		return nil, err
+	}
 
-func NewSchedulerManager(config *rest.Config) *SchedulerManager {
 	manager := SchedulerManager{
-		config:  config,
-		plugins: make(map[string]schedulerinterface.BatchScheduler),
+		rayConfigs: rayConfigs,
+		config:     config,
+		factory:    factory,
+		scheduler:  scheduler,
 	}
-	return &manager
+
+	return &manager, nil
 }
 
 func (batch *SchedulerManager) GetSchedulerForCluster(app *rayv1.RayCluster) (schedulerinterface.BatchScheduler, error) {
-	if schedulerName, ok := app.ObjectMeta.Labels[utils.RaySchedulerName]; ok {
-		return batch.GetScheduler(schedulerName)
-	}
-
-	// no scheduler provided
-	return &schedulerinterface.DefaultBatchScheduler{}, nil
-}
+	batch.Lock()
```

Comment:
why adding the lock here?

Reply:
Double checked the code, seems like there is no risk of running into concurrent issues. Since this will be called in the reconcile logic, which is streamlined. I think we can remove the lock.